programs/fileio.c

```diff
@@ -862,6 +862,193 @@ int FIO_compressFilename(const char* dstFileName, const char* srcFileName,
     return result;
 }
 
+typedef struct {
+    int numActualFrames;
+    int numSkippableFrames;
+    unsigned long long decompressedSize;
+    int canComputeDecompSize;
+    unsigned long long compressedSize;
+    int usesCheck;
+} fileInfo_t;
+
+
+int calcFrameHeaderSize(BYTE frameHeaderDescriptor){
+    int frameContentSizeBytes = 0;
+    int windowDescriptorBytes;
+    int dictionaryIDBytes;
+    int frameContentSizeFlag = frameHeaderDescriptor >> 6;
+    int singleSegmentFlag = (frameHeaderDescriptor & (1 << 5)) >> 5;
+    int dictionaryIDFlag = frameHeaderDescriptor & 3;
+    if(frameContentSizeFlag!=0){
+        frameContentSizeBytes = 1 << frameContentSizeFlag;
+    }
+    else if(singleSegmentFlag){
+        frameContentSizeBytes = 1;
+    }
+    windowDescriptorBytes = singleSegmentFlag ? 0 : 1;
+    dictionaryIDBytes = dictionaryIDFlag ? 1 << (dictionaryIDFlag - 1): 0;
+    return 4 + 1 + windowDescriptorBytes + frameContentSizeBytes + dictionaryIDBytes;
+}
+
+/*
+ * Reads information from file, stores in *info
+ * if successful, returns 0, otherwise returns 1
+ */
+int getFileInfo(fileInfo_t* info, const char* inFileName){
+    FILE* srcFile = FIO_openSrcFile(inFileName);
+    if(srcFile==NULL){
+        return 1;
+    }
+    info->compressedSize = (unsigned long long)UTIL_getFileSize(inFileName);
+    info->decompressedSize = 0;
+    info->numActualFrames = 0;
+    info-> numSkippableFrames = 0;
+    info->canComputeDecompSize = 1;
+    /* begin analyzing frame */
+    while(1){
+        BYTE magicNumberBuffer[4];
+        size_t numBytesRead = fread(magicNumberBuffer, 1, 4, srcFile);
```

Comment:
It might be simpler to unconditionally try to read `ZSTD_FRAMEHEADERSIZE_MAX` bytes.

```c
BYTE headerBuffer[ZSTD_FRAMEHEADERSIZE_MAX];
size_t const numBytesRead = fread(headerBuffer, 1, sizeof(headerBuffer), srcFile);
if (numBytesRead < ZSTD_frameHeaderSize_min) {
    /* handle error */
}
U32 const magicNumber = MEM_readLE32(headerBuffer);

if (magicNumber == ZSTD_MAGICNUMBER) {
    U64 const frameContentSize = ZSTD_getFrameContentSize(headerBuffer, numBytesRead);
    if (frameContentSize == ZSTD_CONTENTSIZE_ERROR || ZSTD_CONTENTSIZE_UNKNOWN) {
        /* handle error */
    }
    size_t const headerSize = ZSTD_frameHeaderSize(headerBuffer, numBytesRead);
    assert(!ZSTD_isError(headerSize));

    int const ret = fseek(srcFile, headerSize, SEEK_SET);
    /* skip the rest of the blocks in the frame */

} else if (magicNumber == ZSTD_MAGIC_SKIPPABLE_START) {
    /* Skippable frames */
}
```